sklearn/decomposition/tests/test_pca.py

```diff
@@ -340,11 +341,35 @@ def test_pca_inverse():
 
 
 def test_pca_validation():
-    X = [[0, 1], [1, 0]]
+    # Ensures that extreme inputs for n_components common to all solvers
+    # (less than 0 or more than the lesser dimension of the input
+    # matrix X) raise errors.
+    X = np.array([[0, 1, 0], [1, 0, 0]])
     for solver in solver_list:
         for n_components in [-1, 3]:
-            assert_raises(ValueError,
-                          PCA(n_components, svd_solver=solver).fit, X)
+            assert_raises_regex(ValueError,
+                                "n_components\=.* must be between .* and min\("
+                                "n_samples, n_features\)\=.* with svd_solver"
+                                "\=\'(?:full|arpack|randomized|auto)\'$",
+                                PCA(n_components, svd_solver=solver).fit, X)
+            # We conduct the same test on X.T so that it is invariant to axis.
+            assert_raises_regex(ValueError,
+                                "n_components\=.* must be between .* and min\("
+                                "n_samples, n_features\)\=.* with svd_solver"
+                                "\=\'(?:full|arpack|randomized|auto)\'$",
+                                PCA(n_components, svd_solver=solver).fit, X.T)
+
+
+def test_n_components_none():
+    # Ensures that n_components == None is handled correctly
+    X = iris.data
+    for solver in solver_list:
+        pca = PCA(svd_solver=solver)
+        pca.fit(X)
+        if solver == 'arpack':
+            assert_equal(pca.n_components_, min(X.shape) - 1)
```

Comment:
Thanks jnothman. As I said, I can add it “alright", but I am not sure of the relevance. 
This particular part tests this code:
```
# Handle n_components==None

if self.n_components is None:
    if self.svd_solver != 'arpack':
        n_components = min(X.shape)
    else:
        n_components = min(X.shape) - 1
else:
    n_components = self.n_components
```
Not only should the axis make no difference (min(X.shape) is also used in the tested code), the assert_equal() function running successfully can only be achieved with the correct n_components, which is all we want.

Do tell me if I'm still missing something.